File: daily_problems/2024/11/1101/personal_submission/cf295c_yawn_sean.py
# Submission link: https://codeforces.com/contest/295/submission/289075491
def main():
    mod = 10 ** 9 + 7
    combs = [[0] * 51 for _ in range(51)]

    for i in range(51):
        combs[i][0] = combs[i][i] = 1
        for j in range(1, i):
            combs[i][j] = combs[i - 1][j] + combs[i - 1][j - 1]
            if combs[i][j] >= mod:
                combs[i][j] -= mod

    n, k = MII()
    k //= 50

    nums = LII()
    c1 = c2 = 0

    for v in nums:
        if v == 50: c1 += 1
        else: c2 += 1

    dp = [[0] * (c2 + 1) for _ in range(c1 + 1)]
    dp[0][0] = 1

    vis = [[0] * (c2 + 1) for _ in range(c1 + 1)]
    vis[0][0] = 1

    ndp = [[0] * (c2 + 1) for _ in range(c1 + 1)]
    nvis = [[0] * (c2 + 1) for _ in range(c1 + 1)]

    for i in range(2 * n + 1):
        for x in range(c1 + 1):
            for y in range(c2 + 1):
                if vis[x][y] == 0: continue
                
                for nx in range(c1 - x + 1):
                    for ny in range(c2 - y + 1):
                        if nx + ny * 2 <= k:
                            ndp[x + nx][y + ny] += dp[x][y] * combs[c1 - x][nx] % mod * combs[c2 - y][ny] % mod
                            nvis[x+ nx][y + ny] = 1
                            if ndp[x + nx][y + ny] >= mod:
                                ndp[x + nx][y + ny] -= mod

        for x in range(c1 + 1):
            for y in range(c2 + 1):
                dp[x][y] = ndp[x][y]
                vis[x][y] = nvis[x][y]
                ndp[x][y] = 0
                nvis[x][y] = 0
        
        if vis[c1][c2]:
            print(2 * i + 1)
            print(dp[c1][c2])
            exit()
        
        for x in range(c1 + 1):
            for y in range(c2 + 1):
                if vis[x][y] == 0: continue
                
                for nx in range(x + 1):
                    for ny in range(y + 1):
                        if nx + ny * 2 <= k and (nx > 0 or ny > 0):
                            ndp[x - nx][y - ny] += dp[x][y] * combs[x][nx] % mod * combs[y][ny] % mod
                            nvis[x - nx][y - ny] = 1
                            if ndp[x - nx][y - ny] >= mod:
                                ndp[x - nx][y - ny] -= mod
        
        for x in range(c1 + 1):
            for y in range(c2 + 1):
                dp[x][y] = ndp[x][y]
                vis[x][y] = nvis[x][y]
                ndp[x][y] = 0
                nvis[x][y] = 0

    print(-1)
    print(0)

File: daily_problems/2024/11/1101/personal_submission/cf295c_yawn_sean.cpp
#include <bits/stdc++.h>
#pragma GCC optimize("O3,Ofast,unroll-loops")
// #include <bits/extc++.h>
// using namespace __gnu_cxx;
// using namespace __gnu_pbds;
using namespace std; 

signed main()
{
    ios_base::sync_with_stdio(false);
    cin.tie(0);
    cout.tie(0);

    int mod = 1e9 + 7;
    vector<vector<int>> combs(51, vector<int>(51, 0));
    for (int i = 0; i <= 50; i ++) {
        combs[i][0] = 1;
        for (int j = 1; j <= i; j ++) {
            combs[i][j] = combs[i - 1][j] + combs[i - 1][j - 1];
            if (combs[i][j] >= mod) combs[i][j] -= mod;
        }
    }

    int n, k;
    cin >> n >> k;
    k /= 50;

    int c1 = 0, c2 = 0;
    for (int i = 0; i < n; i ++) {
        int x;
        cin >> x;
        if (x == 50) c1 ++;
        else c2 ++;
    }

    vector<vector<int>> dp(c1 + 1, vector<int>(c2 + 1, 0));
    vector<vector<int>> ndp(c1 + 1, vector<int>(c2 + 1, 0));
    vector<vector<int>> vis(c1 + 1, vector<int>(c2 + 1, 0));
    vector<vector<int>> nvis(c1 + 1, vector<int>(c2 + 1, 0));

    dp[0][0] = 1;
    vis[0][0] = 1;

    for (int i = 0; i <= 2 * n; i ++) {

        for (int x = 0; x <= c1; x ++) {
            for (int y = 0; y <= c2; y ++) {
                if (!vis[x][y]) continue;

                for (int nx = 0; x + nx <= c1; nx ++) {
                    for (int ny = 0; y + ny <= c2; ny ++) {
                        if (nx + 2 * ny <= k) {
                            ndp[x + nx][y + ny] += 1ll * dp[x][y] * combs[c1 - x][nx] % mod * combs[c2 - y][ny] % mod;
                            nvis[x + nx][y + ny] = 1;
                            if (ndp[x + nx][y + ny] >= mod)
                                ndp[x + nx][y + ny] -= mod;
                        }
                    }
                }
            }
        }

        for (int x = 0; x <= c1; x ++) {
            for (int y = 0; y <= c2; y ++) {
                dp[x][y] = ndp[x][y];
                vis[x][y] = nvis[x][y];
                ndp[x][y] = 0;
                nvis[x][y] = 0;
            }
        }

        if (vis[c1][c2]) {
            cout << 2 * i + 1 << '\n';
            cout << dp[c1][c2];
            return 0;
        }

        for (int x = 0; x <= c1; x ++) {
            for (int y = 0; y <= c2; y ++) {
                if (!vis[x][y]) continue;
                
                for (int nx = 0; nx <= x; nx ++) {
                    for (int ny = 0; ny <= y; ny ++) {
                        if ((nx > 0 || ny > 0) && nx + 2 * ny <= k) {
                            ndp[x - nx][y - ny] += 1ll * dp[x][y] * combs[x][nx] % mod * combs[y][ny] % mod;
                            nvis[x - nx][y - ny] = 1;
                            if (ndp[x - nx][y - ny] >= mod)
                                ndp[x - nx][y - ny] -= mod;
                        }
                    }
                }
            }
        }

        for (int x = 0; x <= c1; x ++) {
            for (int y = 0; y <= c2; y ++) {
                dp[x][y] = ndp[x][y];
                vis[x][y] = nvis[x][y];
                ndp[x][y] = 0;
                nvis[x][y] = 0;
            }
        }
    }

    cout << "-1\n0";

    return 0;
}